### PR TITLE
feature/SOF-7426 feat: displace film minimizing norm of distances

### PR DESCRIPTION
**Pull request by VsevolodX**
- feat: add film/sub norms and optimized displacement

### CODE REVIEW

File: src/py/mat3ra/made/tools/utils/__init__.py [hunk 1]
```diff
@@ -64,6 +65,23 @@ def get_distance_between_coordinates(coordinate1: List[float], coordinate2: List
     return float(np.linalg.norm(np.array(coordinate1) - np.array(coordinate2)))
 
 
+def calculate_norm_of_distances_between_coordinates(coords1: np.ndarray, coords2: np.ndarray) -> float:
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/utils/__init__.py:68]: coords -> coordinates

Comment (timurbazhirov) [reply src/py/mat3ra/made/tools/utils/__init__.py:68]: Should be sum instead of norm

File: src/py/mat3ra/made/tools/utils/__init__.py [hunk 2]
```diff
@@ -64,6 +65,23 @@ def get_distance_between_coordinates(coordinate1: List[float], coordinate2: List
...
+    Args:
+        coords1 (np.ndarray): The first set of coordinates.
+        coords2 (np.ndarray): The second set of coordinates.
+
+    Returns:
+        float: The calculated norm.
+    """
+    tree = cKDTree(coords2)
+    distances, _ = tree.query(coords1)
+    distances = distances[~np.isinf(distances)]
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/utils/__init__.py:81]: Should be removed

File: src/py/mat3ra/made/tools/utils/__init__.py [hunk 3]
```diff
@@ -64,6 +65,23 @@ def get_distance_between_coordinates(coordinate1: List[float], coordinate2: List
...
+
+    Args:
+        coords1 (np.ndarray): The first set of coordinates.
+        coords2 (np.ndarray): The second set of coordinates.
+
+    Returns:
+        float: The calculated norm.
+    """
+    tree = cKDTree(coords2)
+    distances, _ = tree.query(coords1)
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/utils/__init__.py:80]: query doesn't seem to work here, should be just a simple `map(np.linalg.norm)` for the second coords, or consider https://docs.scipy.org/doc/scipy-1.13.1/reference/generated/scipy.spatial.cKDTree.sparse_distance_matrix.html#scipy.spatial.cKDTree.sparse_distance_matrix

File: src/py/mat3ra/made/tools/analyze.py [hunk 4]
```diff
@@ -481,3 +481,44 @@ def get_undercoordinated_atom_indices(
...
+        material (Material): The interface Material object.
+        grid_size (Tuple[int, int]): The size of the grid.
+        search_range (Tuple[float, float]): The range to search for optimal displacements.
+        shadowing_radius (float): The shadowing radius to detect the surface atoms, in Angstroms.
+
+    Returns:
+        List[List[float]]: The optimal displacements.
+    """
+    from .calculate import calculate_norm_of_distances
+    from .modify import displace_interface
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze.py:507]: ???

Comment (timurbazhirov) [reply src/py/mat3ra/made/tools/analyze.py:507]: Remove the inline imports

File: src/py/mat3ra/made/tools/analyze.py [hunk 5]
```diff
@@ -481,3 +481,44 @@ def get_undercoordinated_atom_indices(
     coordination_numbers = get_coordination_numbers(material, indices, cutoff)
     undercoordinated_atoms_indices = [i for i, cn in enumerate(coordination_numbers) if cn <= coordination_threshold]
     return undercoordinated_atoms_indices
+
+
+def get_optimal_displacements(
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze.py:486]: Should be within `interface`?

File: src/py/mat3ra/made/tools/utils/__init__.py (hunk 1 above)
Comment (timurbazhirov) [reply src/py/mat3ra/made/tools/utils/__init__.py:68]: get_norm_of_distances_between_coordinates

File: src/py/mat3ra/made/tools/analyze.py (hunk 5 above)
Comment (timurbazhirov) [reply src/py/mat3ra/made/tools/analyze.py:486]: calculate_optimal_displacement and -> calculate

File: src/py/mat3ra/made/tools/analyze.py [hunk 6]
```diff
@@ -481,3 +481,44 @@ def get_undercoordinated_atom_indices(
...
+    x_values = np.linspace(search_range[0], search_range[1], grid_size[0])
+    y_values = np.linspace(search_range[0], search_range[1], grid_size[1])
+
+    norms = np.zeros(grid_size)
+
+    for i, x in enumerate(x_values):
+        for j, y in enumerate(y_values):
+            displaced_material = displace_interface(material, [x, y, 0], use_cartesian_coordinates=True)
+            norm = calculate_norm_of_distances(displaced_material, shadowing_radius)
+            norms[i, j] = norm
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze.py:518]: The above 3 lines can be separated into a function and `get_optimal_displacements` can be adjusted into a more genering `calculate_on_xy_grid`

File: src/py/mat3ra/made/tools/analyze.py [hunk 7]
```diff
@@ -481,3 +481,44 @@ def get_undercoordinated_atom_indices(
     coordination_numbers = get_coordination_numbers(material, indices, cutoff)
     undercoordinated_atoms_indices = [i for i, cn in enumerate(coordination_numbers) if cn <= coordination_threshold]
     return undercoordinated_atoms_indices
+
+
+def get_optimal_displacements(
+    material: Material,
+    grid_size: Tuple[int, int] = (10, 10),
+    search_range: Tuple[float, float] = (-1.0, 1.0),
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze.py:489]: search_range_x, search_range_y, add units

Comment (timurbazhirov) [reply src/py/mat3ra/made/tools/analyze.py:489]: grid_size_xy = (10, 10)
grid_offset_position_xy = [0, 0]
grid_range_x = (-0.5, 0.5)
grid_range_y = (-0.5, 0.5)
use_cartesian = False

File: src/py/mat3ra/made/tools/utils/__init__.py (hunk 3 above)
Comment (VsevolodX) [reply src/py/mat3ra/made/tools/utils/__init__.py]: ok

File: src/py/mat3ra/made/tools/analyze.py [hunk 8]
```diff
@@ -509,3 +509,57 @@ def get_local_extremum_atom_index(
...
+            modified_material = modifier(
+                material,
+                displacement=[x, y, 0],
+                use_cartesian_coordinates=use_cartesian_coordinates,
+                **modifier_parameters,
+            )
+            result = calculator(modified_material, **calculator_parameters)
+            results_matrix[i, j] = result
+
+    return results_matrix
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze.py:565]: return the matrix of xy_values too

File: src/py/mat3ra/made/tools/build/interface/__init__.py [hunk 9]
```diff
@@ -24,3 +29,36 @@ def create_interface(
     if builder is None:
         builder = SimpleInterfaceBuilder(build_parameters=SimpleInterfaceBuilderParameters())
     return builder.get_material(configuration)
+
+
+def get_optimal_film_displacement(
+    material: Material,
+    grid_size_xy: Tuple[int, int] = (10, 10),
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/interface/__init__.py:36]: calculator should be a parameter

File: src/py/mat3ra/made/tools/calculate.py [hunk 10]
```diff
@@ -125,3 +130,40 @@ def calculate_interfacial_energy(
...
+        film_material, SurfaceTypes.BOTTOM, shadowing_radius=shadowing_radius
+    )
+    substrate_atoms_surface_indices = get_surface_atom_indices(
+        substrate_material, SurfaceTypes.TOP, shadowing_radius=shadowing_radius
+    )
+
+    film_atoms_surface_coordinates = film_material.basis.coordinates
+    film_atoms_surface_coordinates.filter_by_ids(film_atoms_surface_indices)
+    substrate_atoms_surface_coordinates = substrate_material.basis.coordinates
+    substrate_atoms_surface_coordinates.filter_by_ids(substrate_atoms_surface_indices)
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/calculate.py:164]: Should be in interface-related code

File: src/py/mat3ra/made/tools/calculate.py [hunk 11]
```diff
@@ -125,3 +130,40 @@ def calculate_interfacial_energy(
...
+
+    film_coordinates_values = np.array(film_atoms_surface_coordinates.values)
+    substrate_coordinates_values = np.array(substrate_atoms_surface_coordinates.values)
+
+    return get_norm_of_distances_between_coordinates(film_coordinates_values, substrate_coordinates_values)
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/calculate.py:169]: norm_of_distances -> sum_of_inverse_distances_squared here and above

File: src/py/mat3ra/made/tools/calculate.py [hunk 12]
```diff
@@ -125,3 +130,40 @@ def calculate_interfacial_energy(
     surface_energy_film = calculate_surface_energy(film_slab, film_bulk, calculator)
     adhesion_energy = calculate_adhesion_energy(interface, substrate_slab, film_slab, calculator)
     return surface_energy_film + surface_energy_substrate - adhesion_energy
+
+
+@decorator_handle_periodic_boundary_conditions(cutoff=0.25)
+def calculate_norm_of_distances(material: Material, shadowing_radius: float = 2.5) -> float:
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/calculate.py:136]: calculate_film_substrate_interaction_metric, should have `interaction_function` as parameter with `um_of_inverse_distances_squared` as default

File: src/py/mat3ra/made/tools/utils/__init__.py [hunk 13]
```diff
@@ -64,6 +55,28 @@ def get_distance_between_coordinates(coordinate1: List[float], coordinate2: List
     return float(np.linalg.norm(np.array(coordinate1) - np.array(coordinate2)))
 
 
+def get_norm_of_distances_between_coordinates(coordinates_1: np.ndarray, coordinates_2: np.ndarray) -> float:
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/utils/__init__.py:58]: Same as above

File: src/py/mat3ra/made/basis.py [hunk 14]
```diff
@@ -120,8 +120,18 @@ def remove_atom_by_id(self, id=None):
         self.coordinates.remove_item(id)
         self.labels.remove_item(id)
 
-    def filter_atoms_by_ids(self, ids):
+    def filter_atoms_by_ids(self, ids) -> "Basis":
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/basis.py:123]: types for ids and labels below

File: src/py/mat3ra/made/tools/analyze.py [hunk 15]
```diff
@@ -509,3 +509,51 @@ def get_local_extremum_atom_index(
...
+            modified_material = modifier(
+                material,
+                displacement=[x, y, 0],
+                use_cartesian_coordinates=use_cartesian_coordinates,
+                **modifier_parameters,
+            )
+            result = calculator(modified_material, **calculator_parameters)
+            results_matrix[i, j] = result
+
+    return x_values, y_values, results_matrix
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze.py:559]: Let's form a 2d matrix from xy_values as `[x_values, y_values]`

File: src/py/mat3ra/made/tools/calculate.py [hunk 16]
```diff
@@ -125,3 +132,44 @@ def calculate_interfacial_energy(
     surface_energy_film = calculate_surface_energy(film_slab, film_bulk, calculator)
     adhesion_energy = calculate_adhesion_energy(interface, substrate_slab, film_slab, calculator)
     return surface_energy_film + surface_energy_substrate - adhesion_energy
+
+
+class InteractionCalculatorParameters(BaseModel):
+    shadowing_radius: float = 2.5
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/calculate.py:138]: The InteractionCalculatorParameters seems more generic and should not depend on shadowing. But InterfaceInteractionCalculatorParameters could. The parameters should be explained here and then may be referred to on line 152 below

File: src/py/mat3ra/made/tools/modify.py [hunk 17]
```diff
@@ -500,3 +500,15 @@ def displace_interface(
     new_material.to_crystal()
     new_material = wrap_to_unit_cell(new_material)
     return new_material
+
+
+def get_interface_part(
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/modify.py:505]: Maybe analyze would fit better, why modify?

File: src/py/mat3ra/made/tools/analyze.py [hunk 18]
```diff
@@ -509,3 +509,51 @@ def get_local_extremum_atom_index(
         extremum_z_atom = min(z_values, key=lambda item: item[1])
 
     return extremum_z_atom[0]
+
+
+def calculate_on_xy_grid(
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze.py:514]: Should be in calculate or in separate `optimize`

Comment (timurbazhirov) [reply src/py/mat3ra/made/tools/analyze.py:514]: evaluate_calculator_on_xy_grid

File: src/py/mat3ra/made/tools/analyze.py [hunk 19]
```diff
@@ -509,3 +509,51 @@ def get_local_extremum_atom_index(
...
+    material: Material,
+    modifier: Callable,
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze.py:516]: Make safe to be None

File: src/py/mat3ra/made/tools/build/interface/__init__.py [hunk 20]
```diff
@@ -24,3 +29,34 @@ def create_interface(
     if builder is None:
         builder = SimpleInterfaceBuilder(build_parameters=SimpleInterfaceBuilderParameters())
     return builder.get_material(configuration)
+
+
+def get_optimal_film_displacement(
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/interface/__init__.py:34]: Docstring

File: src/py/mat3ra/made/tools/calculate.py [hunk 21]
```diff
@@ -125,3 +132,44 @@ def calculate_interfacial_energy(
...
+    interaction_function: Callable = get_sum_of_inverse_distances_squared
+
+
+@decorator_handle_periodic_boundary_conditions(cutoff=0.25)
+def calculate_film_substrate_interaction_metric(
+    material: Material,
+    shadowing_radius: float = 2.5,
+    interaction_function: Callable = get_sum_of_inverse_distances_squared,
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/calculate.py:146]: Let's create a class `Calculator` with `interaction_function` as parameter and `get_energy(self, material)` as a class

InterfacialCalculator will have shadowing_radius

File: src/py/mat3ra/made/tools/utils/__init__.py [hunk 22]
```diff
@@ -64,6 +55,28 @@ def get_distance_between_coordinates(coordinate1: List[float], coordinate2: List
     return float(np.linalg.norm(np.array(coordinate1) - np.array(coordinate2)))
 
 
+def get_sum_of_inverse_distances_squared(
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/utils/__init__.py:58]: move to `calculate/interaction_functions.py`

File: src/py/mat3ra/made/tools/analyze.py (hunk 18 above)
Comment (timurbazhirov) [reply src/py/mat3ra/made/tools/analyze.py:514]: And isolate to `optimize.py`

File: src/py/mat3ra/made/tools/modify.py (hunk 17 above)
Comment (VsevolodX) [reply src/py/mat3ra/made/tools/modify.py:505]: It returns a material, while analyze means we get some properties from the material.

File: src/py/mat3ra/made/tools/analyze.py [hunk 23]
```diff
@@ -509,3 +509,54 @@ def get_local_extremum_atom_index(
         extremum_z_atom = min(z_values, key=lambda item: item[1])
 
     return extremum_z_atom[0]
+
+
+def evaluate_calculator_on_xy_grid(
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/analyze.py:514]: Should go to optimize

**Review by timurbazhirov — APPROVED** [review]

**Pull request merged**